runtime/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt

```diff
@@ -80,19 +88,32 @@ public open class PluginGeneratedSerialDescriptor(
         @Suppress("DEPRECATION_ERROR")
         if (other !is SerialDescriptor) return false
         if (serialName != other.serialName) return false
-        // TODO compare only serial names
-        if (elementDescriptors() != other.elementDescriptors()) return false
+        if (typeParameterDescriptors != other.typeParameters()) return false
         return true
     }
 
     override fun hashCode(): Int {
         var result = serialName.hashCode()
-        // TODO hashcode only for serial name
-        result = 31 * result + elementDescriptors().hashCode()
+        result = 31 * result + typeParameterDescriptors.hashCode()
         return result
     }
 
+    // todo: should type parameters be in serial name?
     override fun toString(): String {
-        return indices.entries.joinToString(", ", "$serialName(", ")") { it.key + ": " + getElementDescriptor(it.value).serialName }
+        return indices.entries.joinToString(", ", "$serialName(", ")") {
+            it.key + ": " + getElementDescriptor(it.value).serialName
+        }
     }
 }
+
+// this function is for comparing user-defined and plugin-generated descriptors.
+// Do we really need it?
```

Comment:
It's unclear whether we should allow comparing different implementations of SerialDescriptor, but if we do so, equality should be symmetric.

Reply:
I'd say we should not. Asymmetric comparison bugs are really subtle and hard to catch, so we should do our best to avoid it. We don't really want to allow user-defined serializers that are equal to  e.g. `HashSetSerializer`